code/modules/projectiles/gun_system.dm

```diff
@@ -792,28 +687,71 @@ and you're good to go.
 
 ///Wrapper proc to complete the whole firing process.
 /obj/item/weapon/gun/proc/Fire()
-	if(!target || (!gun_user && !istype(loc, /obj/machinery/deployable/mounted/sentry)) || (!CHECK_BITFIELD(flags_item, IS_DEPLOYED) && !able_to_fire(gun_user)))
+	if(!target || (!gun_user && !istype(loc, /obj/machinery/deployable/mounted/sentry)) || (!CHECK_BITFIELD(flags_item, IS_DEPLOYED) && !able_to_fire(gun_user)) || windup_checked == WEAPON_WINDUP_CHECKING)
+		return
+	if(windup_delay && windup_checked == WEAPON_WINDUP_NOT_CHECKED)
+		windup_checked = WEAPON_WINDUP_CHECKING
+		playsound(loc, windup_sound, 30, TRUE)
+		if(!do_after(gun_user, windup_delay, TRUE, src, BUSY_ICON_DANGER, BUSY_ICON_DANGER, ignore_turf_checks = TRUE))
+			windup_checked = WEAPON_WINDUP_NOT_CHECKED
+			return
+		windup_checked = WEAPON_WINDUP_CHECKED
+	if(!target)
+		windup_checked = WEAPON_WINDUP_NOT_CHECKED
 		return
-
 	//The gun should return the bullet that it already loaded from the end cycle of the last Fire().
-	var/obj/projectile/projectile_to_fire = load_into_chamber(gun_user) //Load a bullet in or check for existing one.
-	in_chamber = null //Projectiles live and die fast. It's better to null the reference early so the GC can handle it immediately.
+	var/obj/projectile/projectile_to_fire = in_chamber //Load a bullet in or check for existing one.
 	if(!projectile_to_fire) //If there is nothing to fire, click.
-		click_empty(gun_user)
+		playsound(src, dry_fire_sound, 25, 1, 5)
+		if(CHECK_BITFIELD(reciever_flags, AMMO_RECIEVER_ROTATES_CHAMBER) && !CHECK_BITFIELD(reciever_flags, AMMO_RECIEVER_REQUIRES_UNIQUE_ACTION))
+			cycle(gun_user, FALSE)
+		windup_checked = WEAPON_WINDUP_NOT_CHECKED
 		return
 
 	if(!do_fire(projectile_to_fire))
+		windup_checked = WEAPON_WINDUP_NOT_CHECKED
 		return
-	finish_fire()
+
+	last_fired = world.time
+	SEND_SIGNAL(src, COMSIG_MOB_GUN_FIRED, target, src)
+
+	if(!max_chamber_items)
+		in_chamber = null
+	else
+		if(CHECK_BITFIELD(reciever_flags, AMMO_RECIEVER_TOGGLES_OPEN))
+			casings_to_eject++
+		if(CHECK_BITFIELD(reciever_flags, AMMO_RECIEVER_HANDFULS))
+			QDEL_NULL(in_chamber)
+		else
+			in_chamber = null
+		if(!CHECK_BITFIELD(reciever_flags, AMMO_RECIEVER_REQUIRES_UNIQUE_ACTION))
+			cycle(null)
+		if(length(chamber_items) && CHECK_BITFIELD(reciever_flags, AMMO_RECIEVER_AUTO_EJECT) && CHECK_BITFIELD(reciever_flags, AMMO_RECIEVER_MAGAZINES) && (get_current_rounds(chamber_items[current_chamber_position]) - rounds_per_shot) <= 0)
```

Comment:
this should be just when there's 0 in mag, its currently ejecting with 1 bullet in mag instead of 0 after it actually chambers the round
```suggestion
		if(length(chamber_items) && CHECK_BITFIELD(reciever_flags, AMMO_RECIEVER_AUTO_EJECT) && CHECK_BITFIELD(reciever_flags, AMMO_RECIEVER_MAGAZINES) && (get_current_rounds(chamber_items[current_chamber_position])  < rounds_per_shot)
```
(atleast i think that should solve IMO)